### DISCUSSION TITLE
Create an incubating HTTP implementation using Virtual Threads

### DISCUSSION THREAD

**Discussion by Thihup** [piranhacloud/piranha#2596] [Ideas]
Check if we could create an incubating HTTP implementation using Virtual Threads.

Right now, it is only available for JDK 19+, so we should not make it in the default Piranha package because it would require features that are not in the current LTS (JDK 17).

Maybe in the next LTS, we could use the HTTP implementation using Virtual Threads by default.

(One thing to note is: will we keep using our own HTTP implementation to run/pass the TCK? If so, we'll need to implement the HTTP/2 - and possible HTTP/3 in the future - by hand)

**Comment by mnriem** [comment]
@Thihup I would be fine with you incubating this.

What about ....

1. Create a separate module in the http subdirectory for this work
2. Enable the module by creating a 'jdk19plus' profile that actives for Java 19 or later in the http/pom.xml file
3. Enable Java 19 or later in the build-ea workflow

What do you think?

**Reply by Thihup** [reply]
Sounds great!

**Comment by arjantijms** [comment]
Might also be good to keep an eye out on what Helidon is doing. Dmitry announced something like this on Twitter a short while ago. Maybe we can work together in some way.

**Reply by Thihup** [reply]
I think it is this one, right? https://twitter.com/m0mus/status/1526101284956393472

**Reply by arjantijms** [reply]
That's the very one indeed, good find!

**Reply by Thihup** [reply]
FYI
https://twitter.com/helidon_project/status/1569292011504541697?t=iOJ8kbiCVfoXV6xHARs0vg&s=19

**Reply by arjantijms** [reply]
Wow! Thanks for the pointer!

**Comment by OndroMih** [comment]
This is really interesting indeed. However, I hope that the core Piranha can keep Java 17 as a baseline for a long while, even after the next Java LTS is released. Even with 17 there are still some issues. For example, AWS Lambda doesn't support Java 17 out of the box and a custom runtime is required to run Piranha on AWS Lambda.

**Comment by arjantijms** [comment]
HttpServers are really pluggable in Piranha. We can already run on JDK HttpServer, Grizzly and Netty.

**Reply by OndroMih** [reply]
I know. I just wanted to stress that even if this new HTTP engine requires Java 19+, the rest of Piranha should keep Java 17 as the baseline.

**Reply by mnriem** [reply]
That is why I asked for a jdk19plus Maven profile ;)

**Reply by anbusampath** [reply]
is there any documentation available how to create custom HttpServer for Piranha, so Nima can be customised for that.

**Reply by Thihup** [reply]
You can try out our implementation: https://github.com/piranhacloud/piranha/blob/current/http/virtual/pom.xml

**Reply by anbusampath** [reply]
Looks like latest versions of virtual server not published in maven.Last version I could pull from maven is 23.10.0.
https://mvnrepository.com/artifact/cloud.piranha.http/piranha-http-virtual

**Reply by mnriem** [reply]
Thank you for pointing this out. Note that 24.2.0 made it available again.

**Comment by mnriem** [comment]
FYI https://twitter.com/gunnarmorling/status/1536613918491168769
